src/corelibs/U2Core/src/util/MsaRowUtils.cpp

```diff
@@ -432,4 +432,80 @@ void MsaRowUtils::removeTrailingGapsFromModel(qint64 length, QVector<U2MsaGap> &
     }
 }
 
+QByteArray MsaRowUtils::getGappedSubsequence(const U2Region &region, const QByteArray &sequence, const QVector<U2MsaGap> &gaps) {
+    CHECK(region.length > 0, {});
+    CHECK(!sequence.isEmpty(), QByteArray(region.length, U2Msa::GAP_CHAR));
+    const char *coreSequence = sequence.constData();
+    int coreLength = sequence.length();
+    QByteArray result;
+    result.reserve(region.length);
+    int gapIndex = 0;  // Current gap index.
+    int corePos = 0;  // Current offset in core sequence.
+    int pos = 0;  // Current gapped position in the row.
+    int regionEnd = (int)region.endPos();  // Last (exclusive) global gapped position to include into the result.
```

Comment:
`qint64` to `int` - data loss is possible

Reply:
There is no data loss with 64->32 bit conversion in MSA. We do not support (even open) MSA objects with more than 100M chars in the row. It is a global for UGENE.